src/main/java/ui/components/pickers/RepositoryPickerState.java

```diff
@@ -0,0 +1,143 @@
+        suggestedRepositories.get(nextPosition).setSelected(true);
+    }
+
+    /**
+     * Selects a repository right before the current selected repository in the sorted
+     * suggested repository list. If the current selected repository is the first one,
+     * it will then select the last suggested repository instead.
+     */
+    public void selectPreviousSuggestedRepository() {
+        OptionalInt selectedPositionInSuggested = getSelectedRepositoryPositionInSuggested();
+        assert selectedPositionInSuggested.isPresent();
+        int currentPosition = selectedPositionInSuggested.getAsInt();
+        int nextPosition = currentPosition == 0 ? suggestedRepositories.size() - 1 : currentPosition - 1;
+        suggestedRepositories.get(currentPosition).setSelected(false);
+        suggestedRepositories.get(nextPosition).setSelected(true);
+    }
+
+    /**
+     * Selects a repository in the suggested list. The selected repositoryId must exist
+     * in suggestedRepositories list.
+     */
+    public void setSelectedRepositoryInSuggestedList(String repositoryId) {
+        Optional<PickerRepository> pickerRepository = getPickerRepositoryById(suggestedRepositories, repositoryId);
+        assert pickerRepository.isPresent();
+        clearRepositorySelection();
+        pickerRepository.get().setSelected(true);
+    }
+
+    private OptionalInt getSelectedRepositoryPositionInSuggested() {
+        return IntStream.range(0, suggestedRepositories.size())
+                .filter(index -> suggestedRepositories.get(index).isSelected())
+                .findFirst();
+    }
+
+    /**
+     * Updates List<PickerRepository> suggestedRepositories so that it contains PickerRepositories that match
+     * the user input.
+     *
+     * If there is no existing repository that matches the user's query exactly and the user's query is a well formed
+     * repo id, it will be added to the list of suggested repositories so that the user can pick the new repository.
+     */
+    private void updateSuggestedRepositories(String query) {
+        suggestedRepositories.clear();
+        repositories.stream().forEach(repo -> {
+            if (isMatching(repo, query)) {
+                suggestedRepositories.add(repo);
+            }
+        });
+        Optional<PickerRepository> possibleExactMatch = getPickerRepositoryById(suggestedRepositories, query);
+        if (!possibleExactMatch.isPresent() && !query.isEmpty()) {
+            PickerRepository newRepository = new PickerRepository(query);
+            newRepository.setSelected(true);
+            suggestedRepositories.add(0, newRepository);
+        } else if (!query.isEmpty()) {
+            PickerRepository exactMatch = possibleExactMatch.get();
+            exactMatch.setSelected(true);
+        } else {
+            assert !suggestedRepositories.isEmpty();
+            PickerRepository firstMatching = suggestedRepositories.get(0);
+            firstMatching.setSelected(true);
+        }
```

Comment:
This `if/else` block is a bit convoluted, possibly due to the compound conditionals. Maybe we can try to rearrange it so that the flow more natural, dealing with exceptional cases first and putting attention on the main flow e.g.
```java
if (query.isEmpty()) {
    assert !suggestedRepositories.isEmpty();
    PickerRepository firstMatching = suggestedRepositories.get(0);
    firstMatching.setSelected(true);
    return;
}
if (!possibleExactMatch.isPresent()) {
    PickerRepository newRepository = new PickerRepository(query);
    newRepository.setSelected(true);
    suggestedRepositories.add(0, newRepository);
    return;
}
PickerRepository exactMatch = possibleExactMatch.get();
exactMatch.setSelected(true);